### ISSUE TITLE
Unable to run

### ISSUE THREAD

**Issue by Friggitotti** [ValentinChb/FASTFarm2Simulink#5]
I tried to download and run the test from github but I get different errors:

1. When calling the FastFARM exe in your directory I get the error in the picture attached.

![Screenshot 2024-07-10 202030](https://github.com/ValentinChb/FASTFarm2Simulink/assets/92570491/e76c8b77-b26d-400c-905a-8ff05d3f0c56)

2. When running RunTest_MATLAB alone I get this error:
Ready to connect
Invalid MEX-file 'C:\Users\aless\OneDrive - Politecnico di Milano\Desktop\TESI\FASTFarm2Simulink\SC_MPIServer\MPIServer_Init.mexw64': The specified module could not be found.

Error in SC_MATLAB (line 18)
[nT,nc]=MPIServer_Init(MPI_sharedfile);

ERROR: MATLAB error Exit Status: 0x00000001

Thanks for your time,
Alessandro

**Comment by ValentinChb** [comment]
Hi Alessandro,
Glad to see you're trying to use this. I'll do my best to make it work.
I've never seen the first error but it is purely on the OpenFAST side, this should be fixed by recompiling FastFarm or just using a newer release (that will imply updating the input files). To be clearer, there is no change in FastFarm itself so the official branch from NREL can be used, all mods are in the controller dll.
The second error may be due either to some bad path in Matlab, or to some version incompatibility. Try recompiling the mex file.
BR Valentin

**Comment by Friggitotti** [comment]
Dear BR Valentin,
I'm trying to use your interface for my Master's Thesis.

Regading the first error, may I ask you which version of Fast.Farm did you use?

For the second part, I tried to recompile the mex file in matlab but I got the same error for MPIServer_Init.cpp and MPIServer_Stop.cpp:

>> mex MPIServer_Stop.cpp
Building with 'Microsoft Visual C++ 2022'.
Error using mex
   Creating library MPIServer_Stop.lib and object MPIServer_Stop.exp
MPIServer_Stop.obj : error LNK2019: unresolved external symbol MPIServer_Stop referenced in function "public: virtual void __cdecl
MexFunction::operator()(class matlab::mex::MexIORange<class std::_Vector_iterator<class std::_Vector_val<struct std::_Simple_types<class
matlab::data::Array> > > >,class matlab::mex::MexIORange<class std::_Vector_iterator<class std::_Vector_val<struct std::_Simple_types<class
matlab::data::Array> > > >)"
(??RMexFunction@@UEAAXV?$MexIORange@V?$_Vector_iterator@V?$_Vector_val@U?$_Simple_types@VArray@data@matlab@@@std@@@std@@@std@@@mex@matlab@@0@Z)
MPIServer_Stop.mexw64 : fatal error LNK1120: 1 unresolved externals

Here I attach a picture
![Screenshot 2024-07-12 142607](https://github.com/user-attachments/assets/4db4525e-04de-4770-87b4-ce46f9fc62f7)


Thanks for your help,
Alessandro

**Comment by ValentinChb** [comment]
You need to link the mex file against the MPI server dll, see mex file documentation

**Comment by Friggitotti** [comment]
Which version of FF are you using? I would like to run this with the basic version (no OMP) of the program to have easier life. Do you think is it possible?

**Comment by ValentinChb** [comment]
The version in the GitHub repo is pretty old, 3.0 I guess. As said before, it should be totally fine to update FAST.Farm, as everything is happening in the controller dll. The only thing you need then is to update OpenFAST input files to a newer version. Removing OMP will just call turbines sequantially, that should not change anything beside making it slower. I don't see why that should make your life easier though, unless you want to mess up with the MPI interface.

**Comment by Friggitotti** [comment]
Hi,
I updated the input and solved something. Still I have some issues, first of all I don't understand the exact sequence to start the program, it seems to work by simply click on the run_MATLAB but this procedure seems different then what is written in the readme. 
The new issue I found is this:

Farm_Initialize:T1:Farm_InitFAST:FWrap_Init:FAST_InitializeAll:SrvD_Init:BladedInterface_Init:The
dynamic library .\OpenFAST\T1\ControlData\SCClient_DTUWEC_64.dll could not be loaded. Check that
the file exists in the specified location and that it is compiled for 64-bit applications.
T2:Farm_InitFAST:FWrap_Init:FAST_InitializeAll:SrvD_Init:BladedInterface_Init:The dynamic library
.\OpenFAST\T2\ControlData\SCClient_DTUWEC_64.dll could not be loaded. Check that the file exists
in the specified location and that it is compiled for 64-bit applications.

Thanks for your help.
Alessandro

**Comment by Friggitotti** [comment]
> You need to link the mex file against the MPI server dll, see mex file documentation

I'm having difficulties in understanding which MPI server dll file should I link, in the doc I could only find how to link .lib files but the only one in this folder is not the one needed.

**Comment by ValentinChb** [comment]
Hi,
libSC_MPIServerSubs.a is the (static) library you should link against.
The new issue you are encountering is self explanatory, the WT controller dll file cannot be loaded. Beside the given obvious reasons of path and CPU architecture, it may be because the dll is itself unable to load one of its dependencies (typically the MPI server dll given the original problem). You may use DependencyWalker to confirm this.

**Comment by Friggitotti** [comment]
Hi, 
Thanks for the fast reply. 
I recompiled the cpp files to have the mex64 file but I still have the same error: 

Ready to connect
Invalid MEX-file 'C:\Users\aless\OneDrive - Politecnico di Milano\Desktop\TESI\FASTFarm2Simulink\SC_MPIServer\MPIServer_Init.mexw64': The specified module could not be found.
Error in SC_MATLAB (line 18)
[nT,nc]=MPIServer_Init(MPI_sharedfile);

ERROR: MATLAB error Exit Status: 0x00000001

Then I used DependencyWalker but it finds many many missing dlls for what I understand, I attach a screenshot:
![Screenshot 2024-08-16 130603](https://github.com/user-attachments/assets/48d9328f-604b-473e-bb4e-62ace58d86a7)

Thanks for your help.
Alessandro

**Comment by ValentinChb** [comment]
Hi,
yes DependencyWalker gives a lot of false errors. You've to ignore the system dlls (like kernel32 or msvcrt) and look for any dependency specific to this application. Did you try recompiling the SCClient dll?

**Comment by Friggitotti** [comment]
How can I recompile the SCClient dll? I've found a batch that seems to fit this task but it does not run, it only opens the cmd for 1 ms.

**Comment by Friggitotti** [comment]
Hi,
I solved both issues but I've enountered a new one, from the powershel, when FastFarm starts I get this error:

job aborted:
[ranks] message

[0] fatal error
Fatal error in MPI_Comm_connect: Invalid port, error stack:
MPI_Comm_connect(port="ncalrpc:PCDIALE[18914f95-16ba-428c-b15f-576569c25a7f] ncacn_ip_tcp:PCdiALE[52965] 1", MPI_INFO_NULL, root=0, MPI_COMM_WORLD, newcomm=0x00007FFEAB6E703C) failed
No server is accepting connection on port ncalrpc:PCDIALE[18914f95-16ba-428c-b15f-576569c25a7f] ncacn_ip_tcp:PCdiALE[52965] 1.

Thanks,
Alessandro

**Comment by Friggitotti** [comment]
Hi,
It seems we solved most of the issues.

I have some problems in running the simulation now.
Here the procedure I use:

1. Click on RunTest_MATLAB.bat
2. Wait for "Ready to connect" to appear on one command propt
3. When ready to connect I press a key as written on the other cmd

4. On the first cmd these lines appears:
Ready to connect
Connected
Size avrSWAP: 2x87

5. A PowerShell opens and fastfarm seems to start but then a certain point it blocks:
 [...]
  Heading of the FAST.Farm input file:
   Sample FAST.Farm input file
Running AWAE.
Running InflowWind.
Running WakeDynamics.
Running FASTWrapper.
Running OpenFAST.
OpenFAST input file heading:
    FAST model of the DTU 10MW Reference Wind Turbine on monopile

Running ElastoDyn.
Nodal outputs section of ElastoDyn input file not found or improperly formatted.
Running AeroDyn.
AD15 Nodal Outputs: Nodal output section of AeroDyn input file not found or improperly formatted.
Skipping nodal outputs.
Running InflowWind.
Running ServoDyn.
Running ServoDyn Interface for Bladed Controllers (using Intel Visual Fortran for Windows).
Using legacy Bladed DLL interface.
Running OpenFAST.
OpenFAST input file heading:
    FAST model of the DTU 10MW Reference Wind Turbine on monopile

Running ElastoDyn.
Nodal outputs section of ElastoDyn input file not found or improperly formatted.
Running AeroDyn.
AD15 Nodal Outputs: Nodal output section of AeroDyn input file not found or improperly formatted.
Skipping nodal outputs.
Running InflowWind.
Running ServoDyn.
Running ServoDyn Interface for Bladed Controllers (using Intel Visual Fortran for Windows).
Using legacy Bladed DLL interface.

Farm_Initialize:For efficiency, NumPlanes has been reduced to the number of time steps (67).

I waited some minutes but nothing happens, I tried the other procedure as in the README but that one blocks earlier.

Thanks for your precious help,
Alessandro
